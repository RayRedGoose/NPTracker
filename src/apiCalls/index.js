export const getData = async (type, search) => {
  const searching = (search) ? `?q=${search}`: ''
  const apiKey = 'IpOkxICOi4tJPgJfN3LXvxXuOtWHz5iZdULtD2hd'
  const url = `https://developer.nps.gov/api/v1/${type}${searching}&fields=images&limit=60&api_key=${apiKey}`

  const response = await fetch(url)
  if (!response.ok) {
    throw new Error('There was an error getting parks.')
  }

  const result = await response.json()
  return result.data
}
