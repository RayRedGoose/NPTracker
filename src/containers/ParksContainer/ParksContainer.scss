.parks-content {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  padding: 40px;
  width: 100%;
  overflow: scroll;
}
