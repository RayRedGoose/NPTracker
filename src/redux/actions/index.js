export const addProcessName = name => ({
  type: 'ADD_PROCESS_NAME',
  name
})

export const addProcessStep = step => ({
  type: 'ADD_PROCESS_STEP',
  step
})

export const increaseStep = () => ({
  type: 'INCREASE_STEP'
})

export const addUserInfo = info => ({
  type: 'ADD_USER_INFO',
  info
})

export const addParks = parks => ({
  type: 'ADD_PARKS',
  parks
})

export const changeActiveTab = id => ({
  type: 'CHANGE_ACTIVE_TAB',
  id
})
