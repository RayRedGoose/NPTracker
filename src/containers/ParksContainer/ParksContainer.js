import './ParksContainer.scss'
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { getData } from 'apiCalls'
import { addParks } from 'redux/actions'
import ParkCardPreview from 'containers/ParkCardPreview/ParkCardPreview'

export class ParksContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isLoaded: false,
      error: null
    }
  }

  componentDidMount() {
    if (!this.props.parks.length) this.fetchParks()
  }

  fetchParks = async () => {
    try {
      const parks = await getData('parks', 'park')
      this.setState({isLoaded: true})
      this.props.addParks(parks)
    }
    catch (error) {
      this.setState({
        isLoaded: false,
        error: error.message
      })
    }
  }

  createCards = () => {
    const { parks, type, plannedParks } = this.props
    const planned = parks.filter(park => plannedParks.includes(park.name))
    const data = (type === 'all') ? parks : planned
    return data.map(park => (
      <ParkCardPreview
        key={park.id}
        id={park.id}
        image={park.images[0].url}
        fullName={park.fullName}
        name= {park.name}
        states={park.states}/>
    ))
  }

  render() {
    return (
      <section className="parks-content">
        { this.createCards()}
      </section>
    )
  }
}

export const mapStateToProps = ({parks, plannedParks}) => ({
  parks, plannedParks
})

export const mapDispatchToProps = dispatch => (
  bindActionCreators({
    addParks
  }, dispatch)
)

ParksContainer.propTypes = {
  type: PropTypes.string.isRequired,
  parks: PropTypes.arrayOf(PropTypes.object).isRequired,
  plannedParks: PropTypes.arrayOf(PropTypes.string).isRequired,
  addParks: PropTypes.func.isRequired
}


export default connect(mapStateToProps, mapDispatchToProps)(ParksContainer)
