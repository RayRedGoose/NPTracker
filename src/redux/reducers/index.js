import { combineReducers } from 'redux'
import { process } from './singleReducers/process'
import { user } from './singleReducers/user'

const rootReducer = combineReducers({
  process,
  user
})

export default rootReducer
