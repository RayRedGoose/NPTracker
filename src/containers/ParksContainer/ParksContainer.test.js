import React from 'react'
import { shallow } from 'enzyme'
import { ParksContainer, mapStateToProps, mapDispatchToProps } from './ParksContainer'
import { getData } from 'apiCalls'
import { addParks } from 'redux/actions'

jest.mock('apiCalls')

describe("ParksContainer", () => {
  const mockParksData = [{
    id: 1,
    fullName: 'Some National Park',
    description: 'something about park here',
    images: [{url: 'images/some.jpg'}]
  }]

  describe("ParksContainer container", () => {
    let container, instance
    const addParks = jest.fn()

    beforeEach(() => {
      container = shallow(
        <ParksContainer
          type="all"
          parks={mockParksData}
          plannedParks={[]}
          addParks={addParks} />
      )

      instance = container.instance()
    })

    it("should match snapshot with all data passed in correctly and isLoaded is true", () => {
      container.setState({isLoaded: true})
      expect(container).toMatchSnapshot()
    })

    it("should match snapshot if isLoaded is false but there is no error", () => {
      expect(container).toMatchSnapshot()
    })

    it("should match snapshot if there is an error", () => {
      container.setState({error: 'This is error!'})
      expect(container).toMatchSnapshot()
    })

    it("should call fetchParks method after rendering if there is no parks in props", async () => {
      const container = shallow(
        <ParksContainer
          type="all"
          parks={[]}
          plannedParks={[]}
          addParks={addParks} />
      )

      const instance = container.instance()

      const spy = jest.spyOn(instance, 'fetchParks')
        .mockImplementation(() => Promise.resolve(mockParksData))

      await instance.componentDidMount()

      expect(spy).toHaveBeenCalled()
    })

    it("should call createCards method after rendering if there is park loaded", () => {
      const spy = jest.spyOn(instance, 'createCards')
        .mockImplementation(jest.fn())
      instance.forceUpdate()

      container.setState({isLoaded: true})

      expect(spy).toHaveBeenCalled()
    })

    describe("fetchParks", () => {
      beforeEach(() => {
        getData.mockImplementation(() => Promise.resolve(mockParksData))
      })

      it("should call getData when fetchParks is called", async () => {

        await instance.fetchParks()

        expect(getData).toHaveBeenCalledWith('parks', 'park')
      })

      it("should call addParks prop when fetchParks is called", async () => {

        await instance.fetchParks()

        expect(addParks).toHaveBeenCalledWith(mockParksData)
      })

      it("should change isLoaded state to true", async () => {
        expect(container.state('isLoaded')).toEqual(false)

        await instance.fetchParks()

        expect(container.state('isLoaded')).toEqual(true)
      })

      it("should change isLoaded state to false if fetch is failed", async () => {
        getData.mockImplementation(() => Promise.reject(Error('Failed to fetch')))

        instance.setState({isLoaded: true})

        expect(container.state('isLoaded')).toEqual(true)

        await instance.fetchParks()

        expect(container.state('isLoaded')).toEqual(false)
      })

      it("should change error state if fetch is failed", async () => {
        getData.mockImplementation(() => Promise.reject(Error('Failed to fetch')))

        expect(container.state('error')).toEqual(null)

        await instance.fetchParks()

        expect(container.state('error')).toEqual("Failed to fetch")
      })
    })
  })

  describe("mapStateToProps", () => {
    it("should return the object with correct keys from store", () => {
      const mockStore = {
        parks: mockParksData,
        user: {},
        activeTab: 2
      }

      const expected = {
        parks: mockParksData
      }

      const result = mapStateToProps(mockStore)

      expect(result).toEqual(expected)
    })
  })

  describe("mapDispatchToProps", () => {
    it("should call dispatch with addParks after addParks prop is called", () => {
      const mockDispatch = jest.fn()
      const mockProps = mapDispatchToProps(mockDispatch)
      const actionToDispatch = addParks(mockParksData)

      mockProps.addParks(mockParksData)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})
