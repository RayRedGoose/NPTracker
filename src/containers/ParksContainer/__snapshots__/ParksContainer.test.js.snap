// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParksContainer ParksContainer container should match snapshot with all data passed in correctly 1`] = `
<section
  className="parks-content"
>
  <Connect(ParkCardPreview)
    fullName="Some National Park"
    id={1}
    image="images/some.jpg"
    key="1"
  />
</section>
`;
